include/kernels/clover_outer_product.cuh

```diff
@@ -76,31 +84,35 @@ namespace quda {
       using Spinor = ColorSpinor<typename Arg::real, Arg::nColor, Arg::nSpin>;
       using Link = Matrix<Complex, Arg::nColor>;
 
-      Spinor A = arg.inA(x_cb, 0);
-      Spinor C = arg.inC(x_cb, 0);
-
-#pragma unroll
-      for (int dim=0; dim<4; ++dim) {
-        int shift[4] = {0, 0, 0, 0};
-        shift[dim] = 1;
-        const int nbr_idx = neighborIndex(x_cb, shift, arg.partitioned, arg.parity, arg.X);
-
-        if (nbr_idx >= 0) {
-          Spinor B_shift = arg.inB(nbr_idx, 0);
-          Spinor D_shift = arg.inD(nbr_idx, 0);
-
-          B_shift = (B_shift.project(dim,1)).reconstruct(dim,1);
-          Link result = outerProdSpinTrace(B_shift,A);
-
-          D_shift = (D_shift.project(dim,-1)).reconstruct(dim,-1);
-          result += outerProdSpinTrace(D_shift,C);
-
-          Link temp = arg.force(dim, x_cb, arg.parity);
-          Link U = arg.U(dim, x_cb, arg.parity);
-          result = temp + U*result*arg.coeff;
-          arg.force(dim, x_cb, arg.parity) = result;
-        }
-      } // dim
+      for (int flavor=0; flavor<=arg.doublet; ++flavor){
```

Comment:
Thank you for the review. do you mean something like [b9519f0](https://github.com/lattice/quda/pull/1452/commits/b9519f0314b87ece2dd50fc093a8ea58cabc8996)
for specialization?

Reply:
I meant something a bit simpler: make the `doublet` member of the Arg class a `static constexpr bool` value that is set as a template argument to the Arg class, with it being set differently for the singlet and doublet cases.  This will allow you to use the same functor code for both, with the only difference being a a value of `1` or `2` in the template when instantiating the Arg.  The compiler will then know the trip count of the loop and will be able to unroll it.

Reply:
Yes this is what I was thinking.  😄  Maybe it would be better to have a `n_flavor` static member to make the for loop look more natural, e.g., instead of `<= Arg::doublet`, have `<= Arg::n_flavor`.

Reply:
yes, I change it.

Reply:
[7efe64e](https://github.com/lattice/quda/pull/1452/commits/7efe64eb014b34efd613744cf6bcd6107c396292)